### PR TITLE
Exp 23 change currency

### PR DESCRIPTION
**Pull request by StefanFrusi**

### CODE REVIEW

File: src/main/java/com/endava/expensesmanager/service/impl/ExpenseServiceImpl.java [hunk 1]
```diff
@@ -20,33 +26,51 @@ public ExpenseServiceImpl(ExpenseRepository expenseRepository) {
...
                 .toList();
+
+    }
+
+    public List<Expense> getExpensesByBeginDateAndEndDate(LocalDate beginDate, LocalDate endDate, Integer userId) {
+        return expenseRepository.findExpensesBetweenDatesForUser(beginDate, endDate, userId);
+    }
+
+    @Override
+    public List<Expense> getExpensesByDates(LocalDate beginDate, LocalDate endDate, Integer userId) {
```
Comment (zsoltibv) [inline src/main/java/com/endava/expensesmanager/service/impl/ExpenseServiceImpl.java:65]: this method only calls one method, move impelementation to this method

File: src/main/java/com/endava/expensesmanager/repository/ExpenseRepository.java [hunk 2]
```diff
@@ -1,12 +1,23 @@
...
+import org.springframework.data.repository.query.Param;
 
+import java.time.LocalDate;
 import java.util.List;
-import java.util.Optional;
 
-public interface ExpenseRepository extends JpaRepository<Expense, Integer> {
+public interface ExpenseRepository extends JpaRepository<Expense,Integer> {
+    @Query("SELECT e FROM Expense e WHERE e.expenseDate BETWEEN :startDate AND :endDate AND e.user.userId = :userId")
+    List<Expense> findExpensesBetweenDatesForUser(
```
Comment (zsoltibv) [inline src/main/java/com/endava/expensesmanager/repository/ExpenseRepository.java:13]: you are querying for LocalDate but in db you have LocalDateTime, for me it throws IllegalArgumentException exception.

it works if i change to: 
@Query("SELECT e FROM Expense e WHERE CAST(e.expenseDate AS DATE) BETWEEN :startDate AND :endDate AND e.user.userId = :userId")

**Review by zsoltibv — CHANGES_REQUESTED** [review]

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 3]
```diff
@@ -1,36 +1,72 @@
...
 
         List<ExpenseDto> expenses = expenseService.getExpensesByUserId(userId, startDate, endDate);
         return new ResponseEntity<>(expenses, HttpStatus.OK);
     }
+
+    @GetMapping("/total")
+    public ResponseEntity<?> getTotalSum(@RequestParam int userId, @RequestParam LocalDate startDate, @RequestParam LocalDate endDate, @RequestParam String code) {
+        List<Expense> expenses = currencyService.changeCurrencyTo(code, expenseService.getExpensesByDates(startDate, endDate, userId));
+        BigDecimal sum = BigDecimal.ZERO;
+        for (int i = 0; i < expenses.size(); i++) {
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:47]: This logic should stay in a service, add a new service TotalService where you can use the CurrencyService and EspensesService to build up your logic, beside this we need to have a new controller for Total.

Move this endpoint in a TotalController.

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 4]
```diff
@@ -1,36 +1,72 @@
...
+            sum = sum.add(expenses.get(i).getAmount());
+        }
+        return new ResponseEntity<>(sum, HttpStatus.OK);
+    }
+
+    @GetMapping("/category")
+    public ResponseEntity<?> getTotalSumCategory(@RequestParam int userId, @RequestParam LocalDate startDate, @RequestParam LocalDate endDate, @RequestParam String code) {
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:54]: return type missing. This endpoint should stay in TotalController.

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 5]
```diff
@@ -1,36 +1,72 @@
...
+        if (endDate.isAfter(LocalDate.now()))
+            return new ResponseEntity<>("The end date is invalid", HttpStatus.BAD_REQUEST);
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:56]: throw an error, catch that error in ApiExceptionHandler and return from there a BAD_REQUEST.

@ExceptionHandler(IllegalArgumentException.class) as an example on how to catch it.

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 6]
```diff
@@ -1,36 +1,72 @@
...
+
+        List<Expense> expenses = currencyService.changeCurrencyTo(code, expenseService.getExpensesByDates(startDate, endDate, userId));
+
+        Map<String, BigDecimal> resultMap = expenseService.sortExpenses(expenses);
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:60]: Move logic from controller to service layer.

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 7]
```diff
@@ -1,36 +1,72 @@
...
+
+        return new ResponseEntity<>(resultMap, HttpStatus.OK);
+
+    }
+
+    @GetMapping("/currencies")
+    public ResponseEntity<?> getCurrencies() {
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:67]: We need a new currencies controller.

File: src/main/java/com/endava/expensesmanager/service/CurrencyService.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+import com.endava.expensesmanager.model.entity.Currency;
+import com.endava.expensesmanager.model.entity.Expense;
+
+
+import java.util.List;
+
+public interface CurrencyService {
+    List<Expense> changeCurrencyTo(String code, List<Expense> expenseList);
+     List<Currency> getCurrencies();
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/service/CurrencyService.java:11]: Service should return Dto not the entity.

File: src/main/java/com/endava/expensesmanager/service/ExpenseService.java [hunk 9]
```diff
@@ -2,12 +2,17 @@
...
-
+import java.util.Map;
 
 public interface ExpenseService {
-        Expense addExpense(ExpenseDto expense);
-        List<ExpenseDto> getExpensesByUserId(Integer userId, LocalDate startDate, LocalDate endDate);
+    Expense addExpense(ExpenseDto expense);
+    List<ExpenseDto> getExpensesByUserId(Integer userId, LocalDate startDate, LocalDate endDate);
+    List<Expense> getExpensesByBeginDateAndEndDate(LocalDate beginDate, LocalDate endDate, Integer userId);
+    List<Expense> getExpensesByDates(LocalDate beginDate, LocalDate endDate, Integer userId);
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/service/ExpenseService.java:14]: Service should return Dto not the entity.

File: src/main/java/com/endava/expensesmanager/service/impl/ExpenseServiceImpl.java [hunk 10]
```diff
@@ -20,33 +26,51 @@ public ExpenseServiceImpl(ExpenseRepository expenseRepository) {
...
         }
 
         return expensesStream
                 .map(ExpenseMapper::toDto)
                 .toList();
+
+    }
+
+    public List<Expense> getExpensesByBeginDateAndEndDate(LocalDate beginDate, LocalDate endDate, Integer userId) {
+        return expenseRepository.findExpensesBetweenDatesForUser(beginDate.atStartOfDay(), endDate.atStartOfDay(), userId);
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/service/impl/ExpenseServiceImpl.java:61]: Return Dto, not Entity

File: src/main/java/com/endava/expensesmanager/service/impl/ExpenseServiceImpl.java [hunk 11]
```diff
@@ -20,33 +26,51 @@ public ExpenseServiceImpl(ExpenseRepository expenseRepository) {
...
+    }
+
+    @Override
+    public List<Expense> getExpensesByDates(LocalDate beginDate, LocalDate endDate, Integer userId) {
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/service/impl/ExpenseServiceImpl.java:65]: Return Dto, not Entity

File: src/main/java/com/endava/expensesmanager/service/impl/CurrencyServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+            exchangeRates.put("USD",new BigDecimal("0.22"));
+            exchangeRates.put("RON",new BigDecimal("1"));
+        }
+        System.out.println(exchangeRates);
+    }
+
+    public List<Expense> changeCurrencyTo(String code, List<Expense> expenseList) {
+        List<Expense> newExpenseList = new ArrayList<>();
+
+        for (int i = 0; i < expenseList.size(); i++) {
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/service/impl/CurrencyServiceImpl.java:66]: Use enhance for. It is hard to read the  code as it is, IntelliJ gives some hints regarding the code, check them out.

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 13]
```diff
@@ -1,28 +1,35 @@
 package com.endava.expensesmanager.controller;
 
 import com.endava.expensesmanager.model.dto.ExpenseDto;
+import com.endava.expensesmanager.model.entity.Expense;
+import com.endava.expensesmanager.service.CurrencyService;
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:5]: remove unused imports

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 14]
```diff
@@ -1,28 +1,35 @@
...
 import java.util.List;
+import java.util.Map;
 
 @RestController
 @RequestMapping("/expense")
 @Validated
 public class ExpenseController {
     private final ExpenseService expenseService;
 
+
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:23]: remove extra line

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 15]
```diff
@@ -1,28 +1,35 @@
...
     public ExpenseController(ExpenseService expenseService) {
+
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:25]: remove extra line

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 16]
```diff
@@ -41,9 +48,12 @@ public ResponseEntity<List<ExpenseDto>> getExpensesByUserId(@PathVariable Intege
         return new ResponseEntity<>(expenses, HttpStatus.OK);
     }
 
+
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:51]: remove extra line

File: src/main/java/com/endava/expensesmanager/controller/ExpenseController.java [hunk 17]
```diff
@@ -41,9 +48,12 @@ public ResponseEntity<List<ExpenseDto>> getExpensesByUserId(@PathVariable Intege
...
     @PostMapping("/seed")
     public ResponseEntity<String> seedExpenses(@RequestParam(required = false) Integer nrOfExpenses, @RequestParam(required = false) Integer nrOfDays){
         expenseService.seedExpenses(nrOfExpenses, nrOfDays);
         return new ResponseEntity<>("Expenses added successfully!", HttpStatus.CREATED);
     }
+
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/ExpenseController.java:57]: remove extra line

File: src/main/java/com/endava/expensesmanager/controller/TotalController.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+import com.endava.expensesmanager.model.entity.Expense;
+import com.endava.expensesmanager.service.TotalService;
+import com.endava.expensesmanager.service.impl.TotalServiceImpl;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/TotalController.java:12]: remove unused imports

File: src/main/java/com/endava/expensesmanager/controller/TotalController.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+public class TotalController {
+
+    TotalService totalService;
+    @Autowired
+    public TotalController(TotalService totalService)
+    {
+        this.totalService=totalService;
+    }
+    @GetMapping
+    public ResponseEntity<?> getTotalSum(@RequestParam int userId, @RequestParam LocalDateTime startDate, @RequestParam LocalDateTime endDate, @RequestParam String code) {
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/TotalController.java:31]: add return type, don't use "?"

File: src/main/java/com/endava/expensesmanager/controller/TotalController.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+        return new ResponseEntity<>(totalService.totalExpenseSum(userId,startDate,endDate,code), HttpStatus.OK);
+    }
+    @GetMapping("/category")
+    public ResponseEntity<?> getTotalSumCategory(@RequestParam int userId, @RequestParam LocalDateTime startDate, @RequestParam LocalDateTime endDate, @RequestParam String code) {
+        if (endDate.isAfter(LocalDateTime.now()))
+            return new ResponseEntity<>("The end date is invalid", HttpStatus.BAD_REQUEST);
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/TotalController.java:38]: We can throw IllegalArgumentException in here, make sure that this Exception returns 400 Bad_Request from ApiExceptionHandler

File: src/main/java/com/endava/expensesmanager/controller/TotalController.java [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+
+
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/controller/TotalController.java:40]: delete extra lines

File: src/main/java/com/endava/expensesmanager/service/impl/TotalServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+            sum = sum.add(expens.getAmount());
+        }
+        return sum;
+    }
+
+    public Map<String,BigDecimal> totalExpenseCategory(int userId, LocalDateTime startDate, LocalDateTime endDate, String code)
+    {
+        List<ExpenseDto> expenses = currencyService.changeCurrencyTo(code, expenseService.getExpensesByBeginDateAndEndDate(startDate, endDate, userId));
+
+
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/service/impl/TotalServiceImpl.java:42]: delete extra lines

File: src/main/java/com/endava/expensesmanager/service/impl/TotalServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    private CurrencyService currencyService;
+    private ExpenseService expenseService;
+
+    public TotalServiceImpl(CurrencyService currencyService, ExpenseService expenseService) {
+        this.currencyService = currencyService;
+        this.expenseService = expenseService;
+    }
+
+
+
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/service/impl/TotalServiceImpl.java:27]: delete extra lines

File: src/main/java/com/endava/expensesmanager/service/impl/CurrencyServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+@Value("${exchange-rates-api-enable}")
+    private boolean flag;
+@Value("${exchange-rates-api-key}")
+private String apiKey;
+
+    private WebClient webClient = WebClient.builder().build();
+    @Autowired
+    CurrencyRepository currencyRepository;
+
+
```
Comment (en-alex-trunchi) [inline src/main/java/com/endava/expensesmanager/service/impl/CurrencyServiceImpl.java:36]: delete extra lines

**Review by en-alex-trunchi — CHANGES_REQUESTED** [review]
I didn't have time to check the logic from here CurrencyServiceImpl. @en-dragos-dordea @en-ciprian-parvan do you have some time to check it?

**Review by FitaRoberttt — APPROVED** [review]

**Pull request merged**